### ISSUE TITLE
How to differentiate between signUp and regular ParseUser.save in User class BeforeSave?

### ISSUE THREAD

**Issue by zhoul-HS** [ParsePlatform/parse-server#3104]
Hi!

This is not a bug, but an issue where parse-server behaves differently from Parse.com.
I have anonymous users in my app, and when using ParseUser.signUp (Android SDK) there's an important  difference between parse-server and Parse.com. Parse.com would first check for username conflicts, and if there is one, the beforeSave function wouldn't even be called. This allowed me to make some assumptions in the beforeSave code.
Now in parse-server, the beforeSave is *always* called, and only after I response.success() does it fail to save it.
The problem is that there is some code that shouldn't happen if the internal signUp fails (e.g. duplicate username) but happens anyway since I assume that if beforeSave is called, the username is unique. My solution was to do the check myself via a ParseQuery on the username, but now there's another issue - how do I differentiate between an anonymous user and a new user?

In my app, every new user is automatically saved as an anonymous user. For an anonymous user, there's some things you can't do (change username for example). Now after a while, when he wants to sign up, he enters a username, and in the beforeSave I can't tell if he is trying to change only the username, or is he signing up? If he is trying to sign up, I should allow him to set the username, but if he is just trying to change his username, then I'm suppose to reject the change.

So to summarize: How can I tell in the User class beforeSave if signUp was called or a regular save?

**Comment by hramos** [comment]
Thanks for your question!

We want to make sure to keep signal strong in the GitHub issue tracker – to make sure that it remains the best place to track issues that affect the development of Parse Server.

Questions like yours deserve a purpose-built Q&A forum. Would you like to [post this question to Stack Overflow with the tag #parse.com](http://stackoverflow.com/questions/tagged/parse.com)? We'll be happy to answer there. Post a link to your Stack Overflow question here, so that we don't lose track of it.

You may also use [Server Fault](https://serverfault.com/tags/parse) for questions about managing your own servers.

**Comment by zhoul-HS** [comment]
Totally understandable. here's the link to the question: http://stackoverflow.com/questions/40877666/parse-server-how-to-differentiate-between-signup-and-regular-parseuser-save-in
